asset/boltdb_manager.go

```diff
@@ -0,0 +1,135 @@
+package asset
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+
+	"github.com/boltdb/bolt"
+	"github.com/sensu/sensu-go/types"
+)
+
+var (
+	assetBucketName = []byte("assets")
+)
+
+// A Getter is responsible for fetching (based on fitler selection), verifying,
+// and expanding an asset. Calls to the Get method block until the Asset has
+// fetched, verified, and expanded or it returns an error indicating why getting
+// the asset failed.
+type Getter interface {
+	Get(*types.Asset) (*RuntimeAsset, error)
+}
+
+// BoltDBAssetManager is responsible for the installing and storing the metadata
+// for assets backed by an instance of BoltDB on the local filesystem. BoltDB
+// provides the serialization guarantee that the asset contract specifies.
+// We rely on long-lived BoltDB transactions during Get to provide this
+// mechanism for blocking.
+type BoltDBAssetManager struct {
```

Comment:
I think this type is adding more API surface than is necessary.

BoltDBAssetManager exports all its internals, but that isn't a requirement for any other package right now. I think the LocalStorage and DB fields can be unexported.

Fetcher, Expander and Verifier likely don't need to be anonymously embedded unless there is reason to call these things outside of Get(). I think these things should be unexported also.

Consider unexporting BoltDBAssetManager altogether. If you decide that BoltDBAsset manager should be exported, then it should also have a New function that creates a BoltDBAssetManager with its dependencies as arguments.

Reply:
In that case, I will move the asset_test package into the asset package so that I can manipulate the dependencies of BoltDBAssetManager.

If we're not going to implement a different asset manager, then I think `asset.NewManager()` should be sufficient.

If we want to add an integration test later, we can re-add the asset_test package and test against an actual HTTP resource, but I wanted to avoid that because externally available dependencies, even if s3 is "stable enough."

Reply:
@echlebek -- So I was thinking about this this morning... I see _some_ merit in leaving these things exported in case we want to write a utility for interacting with assets outside of the context of the boltdb manager that's being used in the agent (and eventually backend).

I have thought of a couple of approaches here:

- continue exporting these and allow them to be used piecemeal elsewhere
- unexport them and create a new manager type that doesn't use boltdb so that, for example, the CLI can fetch, verify, and expand an asset

For the latter, however, we may want a command line utility to bypass expanding or we may want to skip verification. There's no mechanism to do that in the interfaces we expose or in the Get interface itself.

We could add an options hash later--or we could do option 1.

Regardless, I am inclined to not take action (potentially requiring unnecessary rework later). If you have a counterargument, I am all ears.

Reply:
I went and thought some more. You're right, re: API surface area. Let's just not give so many options. I'll do what we originally discussed.

Reply:
Unexporting everything makes integration testing this pretty hard without a web server.